core/trino-main/src/main/java/io/trino/operator/scalar/SequenceFunction.java

```diff
@@ -120,6 +120,49 @@ private static Block fixedWidthSequence(long start, long stop, long step, FixedW
         return blockBuilder.build();
     }
 
+    private static int getLength(long start, long stop, long step)
+    {
+        if (step == 1 || step == -1) {
+            if (step == -1) {
+                long temp = start;
+                start = stop;
+                stop = temp;
+            }
+            // 0 start stop
+            if (start >= 0) {
+                if (stop - MAX_RESULT_ENTRIES >= start) {
+                    throw new TrinoException(INVALID_FUNCTION_ARGUMENT, "result of sequence function must not have more than %d entries".formatted(MAX_RESULT_ENTRIES));
+                }
+            }
+            else if (start + MAX_RESULT_ENTRIES <= stop) {
+                throw new TrinoException(INVALID_FUNCTION_ARGUMENT, "result of sequence function must not have more than %d entries".formatted(MAX_RESULT_ENTRIES));
+            }
+
+            return checkMaxEntry(stop - start + 1L);
+        }
+
+        if ((step > 0 && (start >= 0 || stop < 0))
+                || (step < 0 && (start < 0 || stop >= 0))) {
+            return checkMaxEntry((stop - start) / step + 1L);
+        }
+
+        int startLength = abs(checkMaxEntry(start / step));
+        int stopLength = abs(checkMaxEntry(stop / step));
+
+        // 1. step > 0: -step ... startRemain ... 0 ... stopRemain ... step
+        // 2. step < 0: step ... stopRemain ... 0 ... startRemain ... -step
+        long startRemain = start % step;
+        long stopRemain = stop % step;
+        int remainLength;
+        if (step > 0) {
+            remainLength = startRemain + step <= stopRemain ? 2 : 1;
+        }
+        else {
+            remainLength = startRemain + step >= stopRemain ? 2 : 1;
+        }
+        return checkMaxEntry(startLength + stopLength + remainLength);
+    }
```

Comment:
```suggestion
    private static int getLength(long start, long stop, long step)
    {
        // handle the case when start and stop are either both positive, or both negative
        if ((start > 0 && stop > 0) || (start < 0 && stop < 0)) {
            return checkMaxEntry((stop - start) / step + 1);
        }

        // handle small step
        if (step == -1 || step == 1) {
            checkMaxEntry(start);
            checkMaxEntry(stop);
            return checkMaxEntry((stop - start) / step + 1);
        }

        // handle the remaining cases: start and step are of different sign or zero; step absolute value is greater than 1
        int startLength = abs(checkMaxEntry(start / step));
        int stopLength = abs(checkMaxEntry(stop / step));
        long startRemain = start % step;
        long stopRemain = stop % step;
        int remainLength;
        if (step > 0) {
            remainLength = startRemain + step <= stopRemain ? 2 : 1;
        }
        else {
            remainLength = startRemain + step >= stopRemain ? 2 : 1;
        }
        return checkMaxEntry(startLength + stopLength + remainLength);
    }
```

Reply:
@kasiafi Amazing check about the line 132 and 133. The code is much more cleaner than before.
Line 127 still could overflow in the case `(-1, Long.MIN_VALUE, -1)`, we could exclude this.